cmd/kubeadm/app/phases/upgrade/migrate.go

```diff
@@ -0,0 +1,135 @@
+
+func (c CoreDNSCheck) Check() (warnings, errors []error) {
+	if err := c.f(c.client); err != nil {
+		return nil, []error{err}
+	}
+	return nil, nil
+}
+
+// RunCoreDNSMigrationCheck initializes checks related to CoreDNS migration.
+func RunCoreDNSMigrationCheck(client clientset.Interface, ignorePreflightErrors sets.String) error {
+	migrationChecks := []preflight.Checker{
+		&CoreDNSCheck{
+			name:   "CoreDNSUnsupportedPlugins",
+			client: client,
+			f:      checkUnsupportedPlugins,
+		},
+		&CoreDNSCheck{
+			name:   "CoreDNSMigration",
+			client: client,
+			f:      checkMigration,
+		},
+		&CoreDNSCheck{
+			name:   "CoreDNSImage",
+			client: client,
+			f:      checkRelease,
+		},
+	}
+
+	return preflight.RunChecks(migrationChecks, os.Stderr, ignorePreflightErrors)
+}
+
+// checkUnsupportedPlugins checks if there are any plugins included in the current configuration
+// that is unsupported for migration.
+func checkUnsupportedPlugins(client clientset.Interface) error {
+	klog.V(1).Infoln("validating if there are any unsupported CoreDNS plugins in the Config")
+	corefile, currentInstalledCoreDNSversion, err := dns.GetCoreDNSInfo(client)
+	if err != nil {
+		return err
+	}
+	unsupportedCoreDNS, err := migration.Unsupported(currentInstalledCoreDNSversion, kubeadmconstants.CoreDNSVersion, corefile)
+	if err != nil {
+		return err
+	}
+	if unsupportedCoreDNS != nil {
+		var UnsupportedPlugins, UnsupportedVersion string
+		for _, unsup := range unsupportedCoreDNS {
+			UnsupportedPlugins = unsup.Plugin
+			UnsupportedVersion = unsup.Version
+
+		}
+		if UnsupportedPlugins != "" || UnsupportedVersion != "" {
+			return errors.Errorf("there are unsupported plugins in the CoreDNS Configuration")
+		}
+	}
+	return nil
+}
+
+// checkMigration validates if migration of the current CoreDNS ConfigMap is possible.
+func checkMigration(client clientset.Interface) error {
+	klog.V(1).Infoln("validating if migration can be done for the current CoreDNS release.")
+	corefile, currentInstalledCoreDNSversion, err := dns.GetCoreDNSInfo(client)
+	if err != nil {
+		return err
+	}
+
+	_, err = migration.Migrate(currentInstalledCoreDNSversion, kubeadmconstants.CoreDNSVersion, corefile, false)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Checks if the CoreDNS image currently installed is an official released version.
+func checkRelease(client clientset.Interface) error {
+	klog.V(1).Infoln("validating if the image installed is the  official release of CoreDNS.")
+
+	listPods := metav1.ListOptions{LabelSelector: "k8s-app=kube-dns"}
+	pods, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(listPods)
+	if err != nil {
+		return err
+	}
+	coreDNSImageSha := strings.Split(pods.Items[0].Status.ContainerStatuses[0].ImageID, "sha256:")
+
+	isValid := migration.Released(coreDNSImageSha[1])
+	if !isValid {
+		return errors.Errorf("the CoreDNS image currently installed is not an official image. If you still wish to proceed, do so at your own risk.")
```

Comment:
this is not a check we have for the control plane or etcd images.
to my understanding, if someone has built coredns from source at any SHA, `migration.Released` would return `false`. while this is still not recommended, blocking on it with a preflight check might be a bit too strict.

but possibly needs discussion / feedback from more people.

Reply:
+1 for not having this. Users in China may be implicated.

Reply:
I think it is generally safe to remove this check.
  
The main reason for adding this check is to prevent external plugin users from inadvertently  replacing their custom CoreDNS with a stock one. A stock CoreDNS will crash when it encounters an unknown plugin in the Corefile at start up.

But most users just use a stock version of CoreDNS.  And IIUC, those who use CoreDNS compiled with external plugins will upgrade using "phases", and manually upgrade CoreDNS with an updated version compiled with the required external plugins if desired.